dos/src/main.rs

```diff
@@ -269,29 +279,102 @@ fn run_dos_rpc_mode(
     }
 }
 
-fn run_dos(
-    nodes: &[ContactInfo],
+/// Apply given permutation to the vector of items
+fn apply_permutation<'a, T>(permutation: Vec<&usize>, items: &'a Vec<T>) -> Vec<&'a T> {
+    let mut res = Vec::with_capacity(permutation.len());
+    for i in permutation {
+        res.push(&items[*i]);
+    }
+    res
+}
+
+fn run_dos_transactions(
+    target: SocketAddr,
     iterations: usize,
-    payer: Option<&Keypair>,
-    params: DosClientParameters,
+    transaction_params: TransactionParams,
+    rpc_client: &Option<RpcClient>,
 ) {
+    let socket = UdpSocket::bind("0.0.0.0:0").unwrap();
+
+    let num_signatures = transaction_params.num_signatures;
+    let valid_signatures = transaction_params.valid_signatures;
+    let valid_blockhash = transaction_params.valid_blockhash;
+
+    let mut transaction_generator = TransactionGenerator::new(transaction_params);
+
+    // Generate n=1000 unique keypairs, which are used to create
+    // chunks of keypairs.
+    // The number of chunks is described by binomial coefficient
+    // and hence 1000 seems to be a reasonable choice
+    let mut keypairs_flat: Vec<Keypair> = Vec::new();
+    if valid_signatures || valid_blockhash {
+        keypairs_flat = (0..1000 * num_signatures).map(|_| Keypair::new()).collect();
+    }
+
+    let indexes: Vec<usize> = (0..keypairs_flat.len()).collect();
+    let mut it = indexes.iter().permutations(num_signatures);
+    let mut count = 0;
+    let mut total_count = 0;
+    let mut error_count = 0;
+    let mut last_log = Instant::now();
+    loop {
```

Comment:
The code for `mode == rpc` was just extracted by me without much of refactoring, looks like it was not used because it assumed `target == None` which would lead to panic.
Anyways, I simplified the code for this case in this commit https://github.com/solana-labs/solana/pull/24412/commits/d786de66a02589d2eb62448364b1bf484a0cc6f4

But these simplification are not about this particular place. I'm not sure that I fully understood what you mean but I rewrote this code with generic and closures: https://github.com/solana-labs/solana/pull/24412/commits/ff8c04bd0541c31185d02d1ac654e5e96e54b50d